### DISCUSSION TITLE
wagmi next feedback

### DISCUSSION THREAD

**Discussion by tmm** [wevm/wagmi#273] [Announcements]
**Edit: It's out! Thanks for all the feedback everyone.**

Some exciting changes are coming to wagmi: Caching, finite states, abstracted client for Vanilla JS, etc.!

You can download [`wagmi@0.3.0-next.*`](https://www.npmjs.com/package/wagmi/v/0.3.0-next.4) and check it out. Docs are at [next.wagmi.sh](https://next.wagmi.sh). Drop feedback here.

**Comment by zkSoju** [comment]
Ready for it 😄 !

**Comment by flatmapeth** [comment]
I installed and started migrating code over and immediately hit `connector.getSigner is not a function` errors when using `useContractWrite` and `TypeError: client.connector.disconnect is not a function` when calling `disconnect`.

I migrated code according to the examples I found in `examples/playground` and my `_app` looks like this:

```
// Chains for connectors to support
const chains = defaultChains;

// @ts-ignore
const connectors: any = ({ chainId }) => {
  const rpcUrl =
    chains.find((x) => x.id === chainId)?.rpcUrls?.[0] ??
    chain.mainnet.rpcUrls[0];
  return [
    new InjectedConnector({
      chains,
      options: { shimDisconnect: true },
    }),
    new WalletConnectConnector({
      options: {
        infuraId,
        qrcode: true,
      },
    }),
    new CoinbaseWalletConnector({
      options: {
        appName: "App",
        jsonRpcUrl: `${rpcUrl}/${infuraId}`,
      },
    }),
  ];
};

const client = createClient({ autoConnect: false, connectors });

...
App = () => {
  return (
    <Provider client={client}>
      ...
    </Provider>
  )
}
```

Still hacking but though I might post this sooner rather than later. Using `0.3.0-next.4`. Not using `useSigner` anywhere, and wasn't before.

Let me know if you see an obvious derp or if you need more info.

**Reply by jxom** [reply]
are you connected?

**Reply by flatmapeth** [reply]
I thought I was, and everything seemed like it, but I tried manually disconnecting in metamask (now shows `Not connected`) and `useAccount` is still returning my account data:

<img width="773" alt="image" src="https://user-images.githubusercontent.com/93158018/159822149-fedcdca6-3a71-4b2a-87cd-686c8b734d6a.png">

This in itself looks like a bug, maybe unrelated to what I was experiencing before.

**Reply by flatmapeth** [reply]
I cleared localStorage with `localStorage.clear()` in console, reconnected and tried sending a transaction with `useContractWrite` which fails with:

```unsupported provider (argument="provider", value={}, code=INVALID_ARGUMENT, version=providers/5.5.3)```

I'm not manually setting a `provider` anywhere in the codebase.

**Reply by flatmapeth** [reply]
`await account.disconnect()` now works but took quite a bit of time to complete – not sure if that's expected. Definitely can make that nice by using loading states.

**Reply by flatmapeth** [reply]
I'm wondering if I was in some kind of weird state because I was already connected before to `localhost:3000` with wagmi 0.2?

Anyways, `useContractWrite` is still bugging out for me. Anything you spot that's missing in my setup?

**Reply by jxom** [reply]
@flatmapeth – can't really reproduce your issues 😅 can you upload your repo somewhere? or create a [codesandbox](https://codesandbox.io)?

**Reply by eshlox** [reply]
I have the same issue. I'm trying to migrate from `0.2` to `0.3.0-next.4`. I can connect the wallet, use `useAccount`, `useNetwork`, `useBalance` and `useContractRead` without issues, but `useSigner` and `useContractWrite` don't work (everything works in version 0.2).

```js
  const { data } = useSigner();

 // or
 
  const { write } = useContractWrite(
    {
      addressOrName: '0x...',
      contractInterface: 'ABI',
    },
    "approve"
  );
  
  write({args: ['0x...', MaxUint256]})
```

Both calls returns `Error: unsupported provider (argument="provider", value={}, code=INVALID_ARGUMENT, version=providers/5.6.1)`.

You can use the  [playground](https://github.com/tmm/wagmi/tree/next/examples/playground) from examples to reproduce the issue. Just add `useSigner();` to `src/components/Account.tsx` and connect to MetaMask.

<img width="1218" alt="image" src="https://user-images.githubusercontent.com/780763/159995790-32b14dcb-9ccd-42bf-894a-26dff8b0b5b9.png">

**Reply by jxom** [reply]
Can you update to `wagmi@0.3.0-next.5` and try again?

**Reply by eshlox** [reply]
It's even worse now. It seems that every hook returns the same error:

> Error: call revert exception [ See: https://links.ethers.org/v5-errors-CALL_EXCEPTION ] (method="decimals()", errorArgs=null, errorName=null, errorSignature=null, reason=null, code=CALL_EXCEPTION, version=abi/5.6.0)

> Error: call revert exception [ See: https://links.ethers.org/v5-errors-CALL_EXCEPTION ] (method="allowance(address,address)", errorArgs=null, errorName=null, errorSignature=null, reason=null, code=CALL_EXCEPTION, version=abi/5.6.0)

The same problem occurs when I use the example: https://github.com/tmm/wagmi/tree/next/examples/playground

<img width="1021" alt="image" src="https://user-images.githubusercontent.com/780763/160022022-d90558bc-5630-4591-b1b2-2efa0709f905.png">

**Reply by flatmapeth** [reply]
Sorry for the slow turnaround time, and thanks @eshlox for chiming in with runnable example.

@jxom 0.3.0-next.5 fixed the signer error I was getting. Now I'm getting 

```invalid BigNumber value (argument="value", value={"args":[1],"overrides":{"value":{"type":"BigNumber","hex":"0x06f05b59d3b20000"}}}, code=INVALID_ARGUMENT, version=bignumber/5.5.0)```

for a migrated useContractWrite call with the exact same args and overrides as before – the call now looks like this:

```
mint.write({
  args: [numItems],
  overrides: {
    value: utils.parseEther("0.5").mul(numItems),
  },
});
```

**Reply by jxom** [reply]
@flatmapeth – can you update to the latest `next` version and try again? Should work now!

**Reply by jxom** [reply]
@eshlox: I can't reproduce that issue – try reinstall w/ `pnpm install`?

**Reply by flatmapeth** [reply]
@jxom I'm now getting the expected `insufficient funds for intrinsic transaction cost` message – thanks

There's one other thing I've noticed – I have `autoConnect: false` in `createClient` and it's not being respected. It looks like I'm always immediately connected after refreshing the page.

**Reply by jxom** [reply]
@flatmapeth – looks like we were accidently persisting account data (`useAccount` data) when `autoConnect` was falsy. Can you try update now and confirm it's fixed?

**Reply by flatmapeth** [reply]
@jxom it works now – thanks

I plan to port the rest over the weekend and I'll post updates here

**Reply by jxom** [reply]
Thank you for testing!!! 🙏🤝

**Reply by flatmapeth** [reply]
For some good feedback, the API changes are a game changer to me.

It's so easy to make a nice user experience with everything cached.

When there are complex interaction through multiple contracts I just write `const thisItem = useContractRead(...)`, `const thatItem = useContractRead(...)`, `const stake = useContractWrite(...)` and the code just feels great.

I'll have more feedback next week for sure as I complete porting over to 0.3 but I'm loving it so far.

**Reply by eshlox** [reply]
@jxom It works now, the `call revert exception` no longer occurs. 

I noticed that it's not possible to import `useWaitForTransaction` because of missing export: https://github.com/tmm/wagmi/pull/284

**Comment by michaeljohnbennett** [comment]
Are you going to make the provider hooks typed? Would be really good to be able to specify the type other than `BaseProvider` if I wanted to specifically use the Web3 variant to use `.send` etc without resorting to `any` types.

**Reply by tmm** [reply]
Yup! To the degree that this is possible with React Context. If you use the `client` instance instead of the `useProvider` hook, this type will be inferred correctly:

```ts
const client = createClient({
  provider(config) {
    return new providers.AlchemyProvider(config.chainId, 'Your Alchemy ID')
  },
})

// type is AlchemyProvider
client.provider
```

`useProvider` also supports a generic slot if that's more your speed.

**Comment by eshlox** [comment]
It seems that setting `watch` to `true` does not refresh the value (in the previous version, it queried Infura on each new block). I will try to check why it does not work.

```
  const balance = useBalance({
    addressOrName: '0x...',
    watch: true,
  })
```

BTW, is there any reason why the `refetchInterval` is not implemented? IMO it would be a good option to set the refresh interval. Example: we could refresh the token balance every N seconds to decrease the number of queries to Infura. (related: https://github.com/tmm/wagmi/discussions/258)

**Reply by jxom** [reply]
@eshlox – Setting `watch` to true seems to work for me. If you can't get it to work, can you upload your repo or a [codesandbox](https://codesandbox.io/)? Thanks.

Yes, `refetchInterval` is also a good idea!

**Reply by eshlox** [reply]
Maybe it's my fault. 🤔 Everything works when I use the `playground` example. I also noticed that the MetaMask connector does not open the MetaMask window (in my app) when I want to connect. It might be related. I will check it.

Could someone release the newest changes to NPM? The latest version (`0.3.0-next.9`) was released 3 days ago.

**Reply by eshlox** [reply]
OK, I think I know where is the issue. I use Polygon (Mumbai) and Infura. The `watch` option works on all networks except Polygon. When I switch the `playground` example to Infura (the latest version use Alchemy by default) and set Polygon as a default network (it requires a network add-on), it also does not work. What's interesting, it works when I use Alchemy instead of Infura. Maybe it's time to switch to Alchemy. 🤔 

@tmm Would you mind publishing the latest changes to npm? It would be easier to test it.

**Reply by tmm** [reply]
Yup! Will do shortly and let you know.

**Reply by tmm** [reply]
@eshlox [`0.3.0-next.10`](https://www.npmjs.com/package/wagmi/v/0.3.0-next.10) is out

**Comment by tmm** [comment]
Should update `Error` types to allow for accessing ethers-related error info https://github.com/tmm/wagmi/discussions/233

**Comment by eshlox** [comment]
There is a small bug in `useWaitForTransaction` hook: https://github.com/tmm/wagmi/pull/296

**Comment by flatmapeth** [comment]
I'm getting this error, which is probably my fault, but I have no idea where to start looking to fix the problem:

<img width="569" alt="image" src="https://user-images.githubusercontent.com/93158018/162591406-7becb184-26e7-469e-bc00-4c0f4170fe04.png">

Any way to make this clearer for the end user?

**Reply by tmm** [reply]
Yup! Will add a more helpful message. Looks like this occurred because the function you are calling doesn't exist on the contract interface (abi) that is being passed in.

**Reply by flatmapeth** [reply]
I don't think that's possible. I'm not using any new ABI, and all were working before the upgrade.

I'm also seeing whenever this error happens, cache is being busted (`data` reads which were cached start returning undefined).

**Reply by tmm** [reply]
Got it. Will look into more. Any chance you can share this example too? Can [DM](https://twitter.com/awkweb) me if you don't want to share here.

**Reply by flatmapeth** [reply]
I'm trying to pin point this but I'm getting different kind of behavior when using `watch: true` – the error also happens when `watch: false` but it's propagated all the way to the user, resulting in nextjs exception handling (not just logged)

**Reply by flatmapeth** [reply]
Ok I found the problem 😬 It was a method not in the ABI, my bad.

Now the weird thing I'm seeing is (I think) when `watch: true` and the data refresh happens, the data becomes `undefined`, even though I'm using `cacheTime`. Is that expected?

**Reply by flatmapeth** [reply]
The `watch: true` thing is the only out standing issue I'm seeing.

**Reply by tmm** [reply]
Good find. When `watch` is turned on, the `blockNumber` becomes part of the query key so the cache is empty while it fetches new data. Will configure things so the previous data (for the old `blockNumber`and query key) is kept until the new data is fetched.

**Reply by tmm** [reply]
Check out `wagmi@0.3.0-next.15`

**Reply by flatmapeth** [reply]
I think there's still some issue with it – I'm seeing `undefined` value come back when navigating between pages in nextjs. The component which uses the useContractRead is re-mounted (not sure if that matters). I'm doing some more digging.

Edit: I'm pretty sure there's some condition where I switch pages and I had non-undefined value on previous page, I start getting undefined values until next block comes along.

**Reply by flatmapeth** [reply]
I've on rare occasion seen some errors with `REACT_QUERY_OFFLINE_CACHE` being filled. I'm pretty sure I know what it read is, but I don't know why it would be cached since I have `cacheTime: 0`.

**Reply by tmm** [reply]
will look also!

**Reply by flatmapeth** [reply]
Tried to replicate in codesandbox (https://codesandbox.io/s/sweet-grass-gorw92?file=/pages/index.js) but cannot get past ```Error: missing revert data in call exception [ See: https://links.ethers.org/v5-errors-CALL_EXCEPTION ] (code=CALL_EXCEPTION, version=providers/5.6.2)```

Derp, I had the wrong contract address. I've fixed it, now the example shows how it doesn't consistently return from the cache:

<img width="217" alt="image" src="https://user-images.githubusercontent.com/93158018/162839881-a618bdb1-5e75-48e5-99fe-85444fe22d46.png">

Try navigating back and forth between the pages via the "About" link and "Back" link.

**Reply by flatmapeth** [reply]
I added an explicit log with "== null" to see that we're not getting cached values between page transitions:

<img width="293" alt="image" src="https://user-images.githubusercontent.com/93158018/162840117-8449ea73-5a3d-4de6-ad6e-73bedc670863.png">

**Reply by flatmapeth** [reply]
@tmm does the codesandbox help? do you need more information from me?

**Reply by tmm** [reply]
Haven't had a chance to look yet, although it looks like @jxom fixed in 4426c4529a59750835feabcdf6504144c2eacdf5.

**Reply by flatmapeth** [reply]
Awesome, I'm ready to test whenever

**Reply by jxom** [reply]
@flatmapeth – update to .16 and try again? I introduced some changes to improve `useContractRead` caching.

**Reply by flatmapeth** [reply]
@jxom as far as I can tell this version fixes the caching issue I've been having 🎉

**Comment by flatmapeth** [comment]
Is this documentation still valid?

<img width="859" alt="image" src="https://user-images.githubusercontent.com/93158018/162592829-ccbf5c09-7293-4330-b69d-68890e469d57.png">

https://next.wagmi.sh/docs/hooks/useContractRead#configuration-1

Do we still need useMemo?

**Reply by tmm** [reply]
Nope, should be good. Will remove from the docs.

**Comment by bigint** [comment]
Finally moved my app to wagmi 0.3.0 it was smooth, it took around 5 hours for me to migrate cuz a lot of breaking changes, faced some bugs will report it in this thread ASAP 🤗 

Finally, the bundle size became 340kb from 510kb 🚀 

https://gitlab.com/lenster/lenster/-/merge_requests/90

**Comment by bigint** [comment]
Getting this error in `0.3.0-next.16` 🤔 

cc @tmm 

![image](https://user-images.githubusercontent.com/69431456/163087111-0c089cc2-5ff8-47fe-8367-81e21db595b2.png)

**Reply by tmm** [reply]
Can you update to [`0.3.0-next.17`](https://www.npmjs.com/package/wagmi/v/0.3.0-next.17)?

**Reply by bigint** [reply]
@tmm It's still there 😕, I even cleaned node_modules and yarn.lock and installed again, still same issue 🙇🏻 

<img width="1098" alt="image" src="https://user-images.githubusercontent.com/69431456/163912018-76fc9521-50e6-4a85-816e-42084bdea2a4.png">

**Reply by tmm** [reply]
What version of React are you using?

**Reply by bigint** [reply]
@tmm I'm using react 18.0.0

**Reply by tmm** [reply]
@bigint just set up fresh projects with `wagmi@0.3.0-next.17` and `react@18.0.0` with Next.js and Vite, and both worked fine.

Can you share a minimal reproducible example? Also, happy to take a look at the code you are working with (MRE is most useful though).

**Reply by bigint** [reply]
@tmm this is fine now 🙈, it only got fixed after deleting `.next` folder

**Comment by bigint** [comment]
Getting connector not found sometimes 😕  it was fine in 0.2.0

Also difficult to reproduce, its happening once in 100 times

cc @tmm 

![image](https://user-images.githubusercontent.com/69431456/163532911-a2ffc3d6-dd1d-4260-9ceb-88915cd3d28b.png)
![image](https://user-images.githubusercontent.com/69431456/163532916-2c94a278-cf65-4601-a1bd-28068834ec53.png)

**Reply by tmm** [reply]
What wallet are you using when this happens? A code snippet or reproducible example would be helpful!

**Reply by bigint** [reply]
It's causing when metamask wallet is locked, but I figured a way to check it 🙏🏻 

```
if (!activeConnector) {
      disconnect()
    }
```

**Comment by tmm** [comment]
Check out [0.3.0-next.17](https://www.npmjs.com/package/wagmi/v/0.3.0-next.17). Would love to look at some code that you have if this is still happening.

**Reply by albertpurnama** [reply]
dude am I shipping too slow or are you shipping too fast? or both?

sorry for non-discussion reply
